### PR TITLE
Added tests for StrategyBAllocator

### PR DESCRIPTION
**Pull request by spectre-ns**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [ ] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:
This PR adds tests on softmax, MLP (using silu) and SDPA (partial) spanning all LX allocation options. These tests primarily serve as smoke tests for key operations which are important for target model performance. Additionally, tests for specific allocations and core division using co-optimization are checked in `CoOptAllocatorIntegrationTests`. These tests should be updated to remove hard coded values once an accurate model exists to predict expected runtime performance for a given allocation strategy. Currently, these tests are intentionally brittle. Breaking them should serve as a warning of potential performance regression.  

#### Which issue(s) this PR is related to:
N/A

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
No

#### Additional note:
N/A

### CODE REVIEW

File: tests/inductor/test_scratchpad_use.py [hunk 1]
```diff
@@ -146,102 +182,188 @@ def run_test(
...
+                )
+            torch.compiler.reset()
+            with ts_inductor_config.patch(lx_planning=True):
+                result_with_lx, hbm_with_lx = self.measure_hbm_transfers(model, args)
 
         self.assertLess(
             hbm_with_lx,
             hbm_without_lx,
             "Expected LX planning to reduce HBM transfers, but it did not",
         )
```
Comment (jturney) [inline tests/inductor/test_scratchpad_use.py:352]:
```suggestion
        self.assertLess(
            hbm_with_lx,
            hbm_without_lx,
            f"Expected LX planning to reduce HBM transfers, but it did not "
            f"({hbm_with_lx} vs {hbm_without_lx} bytes)",
        )
```

File: tests/inductor/test_scratchpad_use.py [hunk 2]
```diff
@@ -146,102 +182,188 @@ def run_test(
...
+        if not params:
+            return ""
+        return "_".join(
+            [
+                str(params["solver_method"]),
+                f"sc{params['sencores']}",
+                "coopt" if params["co_optimization"] else "nocoopt",
+                "clones" if params["boundary_clones"] else "noclones",
+            ]
+        )
```
Comment (jturney) [inline tests/inductor/test_scratchpad_use.py:258]:
```suggestion
    # How each axis renders into the test-id suffix. Axes not listed fall back to
    # "<name><value>"; this keeps the curated short labels while letting new axes
    # added to ``parameter_axes`` work without editing this method.
    _AXIS_LABELS = {
        "solver_method": lambda v: str(v),
        "sencores": lambda v: f"sc{v}",
        "co_optimization": lambda v: "coopt" if v else "nocoopt",
        "boundary_clones": lambda v: "clones" if v else "noclones",
    }

    @staticmethod
    def _combo_suffix(params: dict) -> str:
        """Readable, test-id-safe suffix for one combo. Empty -> '' (bare name)."""
        if not params:
            return ""
        labels = _ParameterizedScratchpadMeta._AXIS_LABELS
        return "_".join(
            labels[name](value) if name in labels else f"{name}{value}"
            for name, value in params.items()
        )
```

**Comment by spectre-ns** [comment]
@dgrove-oss I'd appreciate a review!

File: tests/inductor/test_scratchpad_use.py [hunk 3]
```diff
@@ -417,5 +546,177 @@ def fn(x):
...
+                "buf0": ("HBM", 2048, (((1, 16),), ((1024, 2),))),
+                "buf1": ("LX", 1048576, (((1024, 32),), ())),
+                "buf2": ("HBM", 1048576, (((1024, 32),), ())),
+                "buf3": ("HBM", 2048, (((1, 16),), ((1024, 2),))),
+                "buf4": ("HBM", 1048576, (((1024, 32),), ())),
+            },
+        )
+
+    def test_mlp_prescribed_allocation(self):
+        """Two-layer linear MLP (``nn.Linear -> silu -> nn.Linear``). With every
```
Comment (isuruf) [inline tests/inductor/test_scratchpad_use.py:669]: If we end up reverting my silu PR because silu opspec is slow, then this will change, but I guess we can update the test.

Comment (spectre-ns) [reply tests/inductor/test_scratchpad_use.py:669]: Happy to apply the reverse before or after this merges. I think this is exactly the sort of thing we want to be sensitive to and how it impacts performance. :) 

Thanks for highlighting this.

**Review by dgrove-oss — COMMENTED** [review]
Some suggestions from code review.  I think things you already know and are ok with, so not mandatory. 

Just let me know if you want to address any of these or if I should merge it as is.

----- 

Summary

Refactors test_scratchpad_use.py to introduce a _ParameterizedScratchpadMeta metaclass that auto-generates a full cartesian-product test sweep across solver methods, core counts, co-optimization, and boundary-clone settings. Adds CoOptAllocatorIntegrationTests with hard-coded allocation fingerprints for softmax, MLP, and SDPA to catch regressions in the co-optimizing allocator's placement decisions.

Review

BLOCKERs

None.

SUGGESTIONs

- [ ] The buf11 gap in test_sdpa_prescribed_allocation (buf10 → buf12 with no buf11) warrants a comment. It's presumably an elided or internally-named buffer, but a future reader will wonder whether it's a typo. Even a one-liner # buf11 is elided by ... would help. (test_scratchpad_use.py, new SDPA block)
- [ ] The CoOptAllocatorIntegrationTests class inherits setUp from BaseTestScratchpadUsage, which patches sencores=1 (or, after the PR, allow_all_ops_in_lx_planning=True). _allocation_fingerprint then patches sencores=32 at the test-case level inside a with block. The ordering (outer setUp patch → inner test-case patch) relies on ts_inductor_config.patch stacking correctly — this is presumably fine, but worth a comment since the sencores that actually governs the compilation is the inner one, not the one from setUp.

QUESTIONs

- [ ] The PR description calls this a cleanup, but the new CoOptAllocatorIntegrationTests tests encode specific hard-coded allocation plans (buffer names, sizes, splits). The description says "these tests are intentionally brittle" — but brittle tests that fail on unrelated allocator improvements can create CI noise. Is there a plan for who owns updating these fingerprints when the allocator improves? A note in the class docstring or a # BRITTLE: marker would help future contributors know to update rather than revert.
- [ ] isuruf's comment flags that reverting the silu opspec PR would change the MLP fingerprint. Has that risk been resolved, or is this PR waiting on that outcome?
- [ ] The @skipUnless(ts_inductor_config.co_optimizing_lx_planning, ...) guard is removed from TestMeasureHBMUsageCoOptimizing. The old guard skipped tests when the feature was off at module import time. Now the tests always run and apply the patch themselves. Is co_optimizing_lx_planning fully production-ready such that running it unconditionally in CI is safe, or is there still a scenario where it's expected to be off?

Verdict

APPROVE — the refactor is clean, CI passes, DCO is signed, and the metaclass approach is a genuine improvement over the previous hand-written solver-variant classes. The three questions above are worth resolving (especially the silu/isuruf concern), but none are blockers. The intentionally-brittle fingerprint tests are a defensible design choice given the explicit callout in the PR description; the main ask is just documenting the buf11 gap and the ownership expectation.

**Comment by spectre-ns** [comment]
@dgrove-oss I have made some clarifying comments in the code but no behavioral changes. Please see below for my detailed response to your review:

#### SUGGESTIONs
1. Added comment to clarify the missing buffer. This is largely peripheral as no part of the repo relies on the buffer names being monotonic. Uniqueness is a requirement and that is preserved.
2. This is a Claude error. In the current diff, the patches are nested and unique such that the same config is not patched multiple times on the same test. All patches are scoped within each test case.

#### QUESTIONs
1. Added a small comment to convey these tests are brittle. @chichun-charlie-liu  is relying on these specific allocation plans to get desired performance. It is important that we are sensitive to regressions moving forward. Long term, I will look to add a more generic metric so we can capture the set of equally performant plans to accommodate other planning strategies.
2. This can be addressed after merge if we decide to revert https://github.com/torch-spyre/torch-spyre/pull/2780 
3. I removed this intentionally as the co-optimizing allocator should be continually tested and serves as a benchmark for future co-optimizing approaches. 

#### SUMMARY
The PR is good to be merged in its current state. This is the basis for future co-optimization testing where we will look to improve the coverage and approach.

**Review by dgrove-oss — APPROVED** [review]
lgtm.

APPROVED — all BLOCKERs from the prior review remain absent, all three explicit prior SUGGESTION/QUESTION items have been addressed in the code or the author's comment. CI is fully green. The two remaining questions are low-risk and can be addressed post-merge. The run_case / @abstractmethod nit is cosmetic.
